File: REL/dgpLinearIV.R
LinearIV <- function(b, n, m){
    
    # Input Args: 
    # b: ture parameter of interests
    # n: sample size, m: number of IVs
    
    # DGP check - July 31, 2017
    
	
	# IV and error terms
	Z <- matrix( rnorm(n*m), n, m );
	Sigma <- sqrtm( matrix( c(0.25, 0.15, 0.15, 0.15, 0.25, 0, 0.15, 0, 0.25), 3, 3 ) );
	E <- matrix( rnorm(n*3), n, 3 ) %*% Sigma;
	e0 <- E[ , 1]; e1 <- E[ , 2]; e2 <- E[ , 3];

	# Structural Equation

	x1 <- Z[ , 1:2] %*% matrix( c(0.5,0.5), 2, 1) + e1;
	x2 <- Z[ , 3:4] %*% matrix( c(0.5,0.5), 2, 1) + e2;
	X <- cbind(x1, x2);

	y <- X %*% b + e0;

	return( list(y = y, X = X, Z = Z) )
}

File: REL/dgpLinearIV.m
function [y, X, Z] = dgpLinearIV(b, n, m)


% n: sample size
% m: number of IVs

%% instruments and disturbances

Z = random('Normal', 0, 1, n, m );
E = random('Normal', 0, 1, n, 3);
Sigma = [0.25, 0.15, 0.15; 0.15, 0.25, 0; 0.15, 0, 0.25];
E = E * Sigma;
e = E(:,1);
v1 = E(:,2);
v2 = E(:,3);

%% structural equation


x1 = Z(:, 1:2) * ones(2, 1) + v1;
x2 = Z(:, 3:4) * ones(2, 1) + v2;

X = [x1, x2];

y = X * b + e;

end
